### PR TITLE
Fix panic in pegged orders when going into auction

### LINKED ISSUE #5825: panic when entering an auction caused by pegged order
# Problem encountered
At some point the market is not in an auction and a pegged order gets repriced. For that it is removed from the book, however it remained in the pegged list. Later when we go into an auction the pegged order list is used to try to get the order from the book and park it but the order does not exist in the book which leads to panic. 

# Expected behaviour
No panic. 

# Logs 
2022-08-08T23:06:45.021414+00:00 n01-d vega[807617]: 2022-08-08T23:06:45.021Z#011PANIC#011execution#011execution/market.go:2378#011Failure to remove order from matching engine#011{"order-id": "0a812db6b86c5ed8eee62478f516ee853114348cc63370bd227fa3c1784f5764", "error": "order does not exist", "errorVerbose": "order does not exist\ncode.vegaprotocol.io/vega/core/matching.init\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/matching/orderbook.go:33\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6222\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:233\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1571"}
2022-08-08T23:06:45.023056+00:00 n01-d vega[807617]: 2022-08-08T23:06:45.022Z#011ERROR#011tendermint#011consensus/state.go:789#011CONSENSUS FAILURE!!!#011{"err": "Failure to remove order from matching engine", "stack": "goroutine 445 [running]:\nruntime/debug.Stack()\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x65\ngithub.com/tendermint/tendermint/internal/consensus.(*State).receiveRoutine.func2()\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/consensus/state.go:789 +0x4c\npanic({0x17ef6c0, 0xc0059ae050})\n\t/usr/local/go/src/runtime/panic.go:838 +0x207\ngo.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc003c7b380, {0xc0049dc100, 0x2, 0x2})\n\t/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.21.0/zapcore/entry.go:232 +0x44c\ngo.uber.org/zap.(*Logger).Panic(0xc0022dbbc0?, {0x1b5da81?, 0x40?}, {0xc0049dc100, 0x2, 0x2})\n\t/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.21.0/logger.go:230 +0x59\ncode.vegaprotocol.io/vega/core/execution.(*Market).parkOrder(0xc0057050e0, {0x1f52c40, 0xc006692750}, {0xc006535ac0, 0x40})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/market.go:2378 +0x27e\ncode.vegaprotocol.io/vega/core/execution.(*Market).parkAllPeggedOrders(0xc0057050e0, {0x1f52c40, 0xc006692750})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/market.go:913 +0xe7\ncode.vegaprotocol.io/vega/core/execution.(*Market).enterAuctionSpecialOrders(0xc0057050e0, {0x1f52c40, 0xc006692750}, {0x2c36788?, 0x0, 0x0})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/special_orders.go:217 +0x66\ncode.vegaprotocol.io/vega/core/execution.(*Market).enterAuction(0xc0057050e0, {0x1f52c40, 0xc006692750})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/market.go:942 +0x56b\ncode.vegaprotocol.io/vega/core/execution.(*Market).checkPriceAndGetTrades(0xc0057050e0, {0x1f52c40, 0xc006692750}, 0xc00809a600)\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/market.go:1474 +0x20e\ncode.vegaprotocol.io/vega/core/execution.(*Market).submitValidatedOrder(0xc0057050e0, {0x1f52c40, 0xc006692750}, 0xc00809a600)\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/market.go:1395 +0xa59\ncode.vegaprotocol.io/vega/core/execution.(*Market).reSubmitPeggedOrders(0xc0057050e0, {0x1f52c40, 0xc006692750}, {0xc00795e300, 0x1f, 0xc00aa8cf40?})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/special_orders.go:101 +0x12d\ncode.vegaprotocol.io/vega/core/execution.(*Market).repriceAllSpecialOrders(0xc0057050e0, {0x1f52c40, 0xc006692750}, 0x0?, {0xc005198000, 0x6, 0x8})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/special_orders.go:178 +0x57f\ncode.vegaprotocol.io/vega/core/execution.(*Market).checkForReferenceMoves(0xc0057050e0, {0x1f52c40, 0xc006692750}, {0xc005198000, 0x6, 0x8}, 0x0)\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/reference_price_moves.go:75 +0x29b\ncode.vegaprotocol.io/vega/core/execution.(*Market).removeExpiredOrders(0xc0057050e0, {0x1f52c40, 0xc006692750}, 0xc006a34e00?)\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/market.go:2989 +0x3aa\ncode.vegaprotocol.io/vega/core/execution.(*Market).OnTick(0xc0057050e0, {0x1f52c40, 0xc006692750}, {0x0?, 0x0?, 0x0?})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/market.go:711 +0x315\ncode.vegaprotocol.io/vega/core/execution.(*Engine).OnTick(0xc00082ad00, {0x1f52c40, 0xc006692750}, {0x1?, 0x1f5af68?, 0x0?})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/execution/engine.go:809 +0x3c7\ncode.vegaprotocol.io/vega/core/vegatime.(*Svc).notify(...)\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/vegatime/service.go:105\ncode.vegaprotocol.io/vega/core/vegatime.(*Svc).SetTimeNow(0xc00080d170, {0x1f52c40, 0xc006692750}, {0x2baa5e0?, 0x17eeb00?, 0x0?})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/vegatime/service.go:78 +0x255\ncode.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(_, {{0xc008702840, 0x20, 0x20}, {{0xb, 0x1}, {0xc0062de520, 0xd}, 0x487, {0x1f273ee1, ...}, ...}, ...})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/processor/abci.go:653 +0x5f2\ncode.vegaprotocol.io/vega/core/blockchain/abci.(*App).BeginBlock(_, {{0xc008702840, 0x20, 0x20}, {{0xb, 0x1}, {0xc0062de520, 0xd}, 0x487, {0x1f273ee1, ...}, ...}, ...})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/core/blockchain/abci/abci.go:62 +0x75\ncode.vegaprotocol.io/vega/cmd/vega/node.(*appW).BeginBlock(_, {{0xc008702840, 0x20, 0x20}, {{0xb, 0x1}, {0xc0062de520, 0xd}, 0x487, {0x1f273ee1, ...}, ...}, ...})\n\t/jenkins/workspace/private/Deployments/Veganet/Devnet/vega/cmd/vega/node/app_wrapper.go:68 +0x55\ngithub.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(_, {_, _}, {{0xc008702840, 0x20, 0x20}, {{0xb, 0x1}, {0xc0062de520, 0xd}, ...}, ...})\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/abci/client/local_client.go:285 +0x118\ngithub.com/tendermint/tendermint/internal/proxy.(*appConnConsensus).BeginBlockSync(_, {_, _}, {{0xc008702840, 0x20, 0x20}, {{0xb, 0x1}, {0xc0062de520, 0xd}, ...}, ...})\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/proxy/app_conn.go:93 +0x14d\ngithub.com/tendermint/tendermint/internal/state.execBlockOnProxyApp({0x1f51d60?, 0xc0000139c0}, {0x1f58690, 0xc000f2c768}, 0xc0067e8960, {0x1f5be48, 0xc000495c50}, 0x486?)\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/state/execution.go:343 +0x3ec\ngithub.com/tendermint/tendermint/internal/state.(*BlockExecutor).ApplyBlock(_, {{{0xb, 0x1}, {0xc00234adfa, 0x6}}, {0xc00234ae00, 0xd}, 0x1, 0x486, {{0xc00944f8e0, ...}, ...}, ...}, ...)\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/state/execution.go:161 +0x151\ngithub.com/tendermint/tendermint/internal/consensus.(*State).finalizeCommit(0xc0023dc000, 0x487)\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/consensus/state.go:1722 +0x9a5\ngithub.com/tendermint/tendermint/internal/consensus.(*State).tryFinalizeCommit(0xc0023dc000, 0x487)\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/consensus/state.go:1636 +0x305\ngithub.com/tendermint/tendermint/internal/consensus.(*State).enterCommit.func1()\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/consensus/state.go:1570 +0xa5\ngithub.com/tendermint/tendermint/internal/consensus.(*State).enterCommit(0xc0023dc000, 0x487, 0x0)\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/consensus/state.go:1609 +0xbea\ngithub.com/tendermint/tendermint/internal/consensus.(*State).addVote(0xc0023dc000, 0xc000fef540, {0xc003b632c0, 0x28})\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/consensus/state.go:2204 +0xb9c\ngithub.com/tendermint/tendermint/internal/consensus.(*State).tryAddVote(0xc0023dc000, 0xc000fef540, {0xc003b632c0?, 0xc004578e00?})\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/consensus/state.go:2002 +0x2c\ngithub.com/tendermint/tendermint/internal/consensus.(*State).handleMsg(0xc0023dc000, {{0x1f429c0?, 0xc007b61b48?}, {0xc003b632c0?, 0x0?}})\n\t/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/tendermint@v0.35.10-0.20220724091544-f328f3ea62c3/internal/consensu

### PR DESCRIPTION
**Pull request by ze97286**
Closes #5825

### CODE REVIEW

**Review by jeremyletang — DISMISSED** [review]

File: core/matching/orderbook.go [hunk 1]
```diff
@@ -1056,8 +1056,8 @@ func (b *OrderBook) Settled() []*types.Order {
 // GetAllUnParkedPeggedOrderIDs returns the order identifiers of all pegged orders in the order book that are not parked.
 func (b *OrderBook) GetAllUnParkedPeggedOrderIDs() []string {
 	pegged := make([]string, 0, len(b.peggedOrders))
-	for _, o := range b.peggedOrders {
-		if o.Status != vega.Order_STATUS_PARKED {
+	for ID := range b.peggedOrders {
+		if o, ok := b.ordersByID[ID]; ok && o.Status != vega.Order_STATUS_PARKED {
```
Comment (jeremyletang) [inline core/matching/orderbook.go:1060]: Well if this is not true, we should probably panic here right? No parked orders should be on the book?

**Review by jeremyletang — DISMISSED** [review]

File: core/matching/orderbook.go [hunk 2]
```diff
@@ -1034,8 +1046,24 @@ func (b *OrderBook) Settled() []*types.Order {
...
 	return orders
 }
+
+// GetActivePeggedOrderIDs returns the order identifiers of all pegged orders in the order book that are not parked.
+func (b *OrderBook) GetActivePeggedOrderIDs() []string {
+	pegged := make([]string, 0, len(b.peggedOrders))
+	for ID := range b.peggedOrders {
+		if o, ok := b.ordersByID[ID]; ok {
+			if o.Status == vega.Order_STATUS_PARKED {
+				panic("unexpected parked pegged order in order book")
```
Comment (jeremyletang) [inline core/matching/orderbook.go:1062]:
```suggestion
				m.log.Panic("unexpected parked pegged order in order book", logging.Order(o))
```

**Review by wwestgarth — DISMISSED** [review]

File: core/execution/pegged_orders.go [hunk 3]
```diff
@@ -194,22 +163,14 @@ func (p *PeggedOrders) RemoveAllParkedForParty(
...
-	for k := range p.orders {
-		if _, parked := p.isParked[k]; !parked {
-			orders = append(orders, k)
-		}
-	}
-	sort.Strings(orders)
-	return
-}
+func (p *PeggedOrders) GetParkedIDs() []string {
+	// the reason for this is that if an pegged order cannot be priced it remains in the book as parked in which case it will be both in the
```
[the remark this answers is not in the document]
Comment (ze97286) [reply core/execution/pegged_orders.go:167]: indeed, the comment doesn't make sense anymore. I'll remove it.

File: core/execution/event_generation_test.go [hunk 4]
```diff
@@ -422,7 +422,7 @@ func TestEvents_PeggedOrderNotAbleToRepriceDueToMargin(t *testing.T) {
 	require.NotNil(t, o5conf)
 	require.NoError(t, err)
 
-	// Move the best bid price up so that the pegged order cannot reprice
+	// Move the best bid price up so that the pegged order cannot repriced
```
Comment (peterbarrow) [inline core/execution/event_generation_test.go:425]:
```suggestion
	// Move the best bid price up so that the pegged order cannot be repriced
```

**Review by wwestgarth — DISMISSED** [review]

**Review by peterbarrow — DISMISSED** [review]

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**